packages/neutrino/Neutrino.js

```diff
@@ -138,38 +142,46 @@ module.exports = class Neutrino {
 
   bindMainsOnOptions(options, optionsSource) {
     Object
-      .keys(options.mains)
-      .forEach(key => {
-        let value = options.mains[key];
+      .entries(options.mains)
+      .forEach(([key, value]) => {
+        let normalizedConfig = normalizeMainConfig(value);
 
         Reflect.defineProperty(options.mains, key, {
           enumerable: true,
           get() {
             const source = optionsSource &&
               optionsSource.source || options.source;
 
-            return normalizePath(source, value);
+            return {
+              ...normalizedConfig,
+              // Lazily normalise the path, in case `source` changes after mains is updated.
+              entry: normalizePath(source, normalizedConfig.entry)
+            };
           },
           set(newValue) {
-            value = newValue;
+            normalizedConfig = normalizeMainConfig(newValue);
           }
         });
       });
 
     this.mainsProxy = new Proxy(options.mains, {
```

Comment:
`mainsProxy` appears to be unused. Do we still need it?

Reply:
We need the Proxy itself, but only exposed it on the Neutrino API just in case. We can certainly remove it if we don't foresee exposing it as being useful.